File: sources/Deltares.Probabilistic.PWrapper/stochast.py
from utils import *
import interface

class Stochast:

	def __init__(self, id = None):
		if id is None:
			self._id = interface.Create('stochast')
		else:
			self._id = id

		self._clear_values()

	def __del__(self):
	    interface.Destroy(self._id)

	@property   
	def name(self):
		return self._name
		
	@property   
	def fullname(self):
		return self._fullname
		
	@property   
	def submodel(self):
		return self._submodel
		
	def _clear_values(self):
		self._distribution = None
		self._inverted = None
		self._truncated = None
		self._mean = None
		self._deviation = None
		self._variation = None
		self._shift = None
		self._shift_b = None
		self._minimum = None
		self._maximum = None
		self._rate = None
		self._shape = None
		self._shape_b = None
		self._scale = None
		self._location = None
		self._observations = None
		self._design_fraction = None
		self._design_factor = None
		self._discrete_values = EventList(self._discrete_values_changed)
		self._histogram_values = EventList(self._histogram_values_changed)
		self._fragility_values = EventList(self._fragility_values_changed)

	def _clear_derived_values(self):
		self._distribution = None
		self._mean = None
		self._deviation = None
		self._variation = None

	def clear(self):
		interface.ClearVariable(self._fullname)
		self._clear_values()

	@property   
	def distribution(self):
		if self._distribution is None:
			self._distribution = interface.GetStringValue(self._id, 'distribution')
		return self._distribution
		
	@distribution.setter
	def distribution(self, value):
		self._clear_values();
		self._distribution = value
		interface.SetStringValue(self._id, 'distribution', value)

	@property   
	def inverted(self):
		if self._inverted is None:
			self._inverted = interface.GetBoolValue(self._id, 'inverted')
		return self._inverted
		
	@inverted.setter
	def inverted(self, value):
		self._inverted = value
		self._clear_derived_values();
		interface.SetBoolValue(self._id, 'inverted', value)

	@property   
	def truncated(self):
		if self._truncated is None:
			self._truncated = interface.GetBoolValue(self._id, 'truncated')
		return self._truncated
		
	@truncated.setter
	def truncated(self, value):
		self._truncated = value
		self._clear_derived_values();
		interface.SetBoolValue(self._id, 'truncated', value)

	@property   
	def mean(self):
		if self._mean is None:
			self._mean = interface.GetValue(self._id, 'mean')
		return self._mean
		
	@mean.setter
	def mean(self, value):
		self._clear_values();
		interface.SetValue(self._id, 'mean', value)

	@property   
	def deviation(self):
		if self._deviation is None:
			self._deviation = interface.GetValue(self._id, 'deviation')
		return self._deviation
		
	@deviation.setter
	def deviation(self, value):
		self._clear_values();
		interface.SetValue(self._id, 'deviation', value)
		
	@property   
	def variation(self):
		if self._variation is None:
			self._variation = interface.GetValue(self._id, 'variation')
		return self._variation
		
	@variation.setter
	def variation(self, value):
		self._clear_values();
		interface.SetValue(self._id, 'variation', value)

	@property   
	def location(self):
		if self._location is None:
			self._location = interface.GetValue(self._id, 'location')
		return self._location
		
	@location.setter
	def location(self, value):
		self._location = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'location', value)

	@property   
	def scale(self):
		if self._scale is None:
			self._scale = interface.GetValue(self._id, 'scale')
		return self._scale
		
	@scale.setter
	def scale(self, value):
		self._scale = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'scale', value)

	@property   
	def shift(self):
		if self._shift is None:
			self._shift = interface.GetValue(self._id, 'shift')
		return self._shift
		
	@shift.setter
	def shift(self, value):
		self._shift = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'shift', value)

	@property   
	def shift_b(self):
		if self._shift_b is None:
			self._shift_b = interface.GetValue(self._id, 'shift_b')
		return self._shift_b
		
	@shift_b.setter
	def shift_b(self, value):
		self._shift_b = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'shift_b', value)

	@property   
	def minimum(self):
		if self._minimum is None:
			self._minimum = interface.GetValue(self._id, 'minimum')
		return self._minimum
		
	@minimum.setter
	def minimum(self, value):
		self._minimum = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'minimum', value)

	@property   
	def maximum(self):
		if self._maximum is None:
			self._maximum = interface.GetValue(self._id, 'maximum')
		return self._maximum
		
	@maximum.setter
	def maximum(self, value):
		self._maximum = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'maximum', value)

	@property   
	def shape(self):
		if self._shape is None:
			self._shape = interface.GetValue(self._id, 'shape')
		return self._shape
		
	@shape.setter
	def shape(self, value):
		self._shape = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'shape', value)

	@property   
	def shape_b(self):
		if self._shape_b is None:
			self._shape_b = interface.GetValue(self._id, 'shape_b')
		return self._shape_b
		
	@shape_b.setter
	def shape_b(self, value):
		self._shape_b = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'shape_b', value)

	@property   
	def rate(self):
		if self._rate is None:
			self._rate = interface.GetValue(self._id, 'rate')
		return self._rate
		
	@rate.setter
	def rate(self, value):
		self._rate = value
		interface.SetValue(self._id, 'rate', value)

	@property   
	def observations(self):
		if self._observations is None:
			self._observations = interface.GetValue(self._id, 'observations')
		return self._observations
		
	@observations.setter
	def observations(self, value):
		self._observations = value
		self._clear_derived_values();
		interface.SetValue(self._id, 'observations', value)

	@property   
	def discrete_values(self):
		return self._discrete_values

	def _discrete_values_changed(self):
		self._clear_derived_values()
		values = [discrete_value._id for discrete_value in self._discrete_values]
		interface.SetArrayValue(self._id, 'discrete_values', values)

	@property   
	def histogram_values(self):
		return self._histogram_values

	def _histogram_values_changed(self):
		self._clear_derived_values()
		values = [histogram_value._id for histogram_value in self._histogram_values]
		interface.SetArrayValue(self._id, 'histogram_values', values)

	@property   
	def fragility_values(self):
		return self._fragility_values

	def _fragility_values_changed(self):
		self._clear_derived_values()
		values = [fragility_value._id for fragility_value in self._fragility_values]
		interface.SetArrayValue(self._id, 'fragility_values', values)

	@property   
	def design_fraction(self):
		if self._design_fraction is None:
			self._design_fraction = interface.GetValue(self._id, 'design_fraction')
		return self._design_fraction
		
	@design_fraction.setter
	def design_fraction(self, value):
		self._design_fraction = value
		interface.SetValue(self._id, 'design_fraction', value)

	@property   
	def design_factor(self):
		if self._design_factor is None:
			self._design_factor = interface.GetValue(self._id, 'design_factor')
		return self._design_factor
		
	@design_factor.setter
	def design_factor(self, value):
		self._design_factor = value
		interface.SetValue(self._id, 'design_factor',  value)
		
	@property   
	def design_value(self):
		return interface.GetValue(self._id, 'design_value')
		
	@design_value.setter
	def design_value(self, value):
		interface.SetValue(self._id, 'design_value', value)

	def get_quantile(self, quantile):
		return interface.GetArgValue(self._id, 'quantile', quantile)
		
	def get_x_from_u(self, u):
		return interface.GetArgValue(self._id, 'x_from_u', u)
		
	def get_u_from_x(self, x):
		return interface.GetArgValue(self._id, 'u_from_x', x)
		
	def get_correlation(self, other):
		if type(other) is Stochast:
			other = other.name
		return interface.GetCorrelation(self._fullname, other)
		  
	def set_correlation(self, other, correlation):
		if type(other) is Stochast:
			other = other.name
		interface.SetCorrelation(self._fullname, other, correlation)


class DiscreteValue:

	def __init__(self):
		self._id = interface.Create('discrete_value')
  
	@property   
	def x(self):
		return interface.GetValue(self._id, 'x')
		
	@x.setter
	def x(self, value):
		interface.SetValue(self._id, 'x',  value)
		
	@property   
	def amount(self):
		return interface.GetValue(self._id, 'amount')
		
	@amount.setter
	def amount(self, value):
		interface.SetValue(self._id, 'amount',  value)

class FragilityValue:

	def __init__(self):
		self._id = interface.Create('fragility_value')
		
	@property   
	def x(self):
		return interface.GetValue(self._id, 'x')
		
	@x.setter
	def x(self, value):
		interface.SetValue(self._id, 'x',  value)
		
	@property   
	def reliability_index(self):
		return interface.GetValue(self._id, 'reliability_index')
		
	@reliability_index.setter
	def reliability_index(self, value):
		interface.SetValue(self._id, 'reliability_index',  value)

	@property   
	def probability_of_failure(self):
		return interface.GetValue(self._id, 'probability_of_failure')
		
	@probability_of_failure.setter
	def probability_of_failure(self, value):
		interface.SetValue(self._id, 'probability_of_failure',  value)
		
	@property   
	def probability_of_non_failure(self):
		return interface.GetValue(self._id, 'probability_of_non_failure')
		
	@probability_of_non_failure.setter
	def probability_of_non_failure(self, value):
		interface.SetValue(self._id, 'probability_of_non_failure',  value)
		
	@property   
	def return_period(self):
		return interface.GetValue(self._id, 'return_period')
		
	@return_period.setter
	def return_period(self, value):
		interface.SetValue(self._id, 'return_period',  value)

class HistogramValue:

	def __init__(self):
		self._id = interface.Create('histogram_value')
  
	@property   
	def lower_bound(self):
		return interface.GetValue(self._id, 'lower_bound')
		
	@lower_bound.setter
	def lower_bound(self, value):
		interface.SetValue(self._id, 'lower_bound',  value)
		
	@property   
	def upper_bound(self):
		return interface.GetValue(self._id, 'upper_bound')
		
	@upper_bound.setter
	def upper_bound(self, value):
		interface.SetValue(self._id, 'upper_bound',  value)
		
	@property   
	def amount(self):
		return interface.GetValue(self._id, 'amount')
		
	@amount.setter
	def amount(self, value):
		interface.SetValue(self._id, 'amount',  value)

File: sources/Deltares.Probabilistic.NetCS/Statistics/Stochast.cs
// Copyright (C) Stichting Deltares. All rights reserved.
//
// This file is part of the Probabilistic Library.
//
// The Probabilistic Library is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.
//
using System.Collections.Generic;
using System.Linq;
using Deltares.Probabilistic.Utils;

namespace Deltares.Probabilistic.Statistics
{
    public class Stochast
    {
        private int id = 0;
        private string name = string.Empty;
        private bool nameAssigned = false;

        private CallBackList<HistogramValue> histogramValues = null;
        private CallBackList<DiscreteValue> discreteValues = null;
        private CallBackList<FragilityValue> fragilityValues = null;
        private CallBackList<ContributingStochast> contributingStochasts = null;
        private CallBackList<ConditionalValue> conditionalValues = null;
        private Stochast conditionalSource = null;

        public Stochast()
        {
            this.id = Interface.Create("stochast");
            ObjectFactory.Register(this, id);
        }

        public Stochast(int id)
        {
            this.id = id;
            ObjectFactory.Register(this, id);
        }

        ~Stochast()
        {
            ObjectFactory.Destroy(id);
        }

        public int GetId()
        {
            return id;
        }

        internal void SetId(int id)
        {
            this.id = id;
        }

        public string Name
        {
            get
            {
                if (!nameAssigned)
                {
                    name = Interface.GetStringValue(id, "name");
                    nameAssigned = true;
                }

                return name;
            }
            set
            {
                Interface.SetStringValue(id, "name", value);
                name = value;
                nameAssigned = true;
            }
        }

        public DistributionType DistributionType
        {
            get { return DistributionTypeConverter.ConvertFromString(Interface.GetStringValue(id, "distribution")); }
            set { Interface.SetStringValue(id, "distribution", DistributionTypeConverter.ConvertToString(value)); }
        }

        public ConstantParameterType ConstantParameterType
        {
            get { return ConstantParameterTypeConverter.ConvertFromString(Interface.GetStringValue(id, "constant_parameter")); }
            set { Interface.SetStringValue(id, "constant_parameter", ConstantParameterTypeConverter.ConvertToString(value)); }
        }

        public bool Inverted
        {
            get { return Interface.GetBoolValue(id, "inverted"); }
            set { Interface.SetBoolValue(id, "inverted", value); }
        }

        public bool Truncated
        {
            get { return Interface.GetBoolValue(id, "truncated"); }
            set { Interface.SetBoolValue(id, "truncated", value); }
        }

        public double Mean
        {
            get { return Interface.GetValue(id, "mean"); }
            set { Interface.SetValue(id, "mean", value); }
        }

        public double Deviation
        {
            get { return Interface.GetValue(id, "deviation"); }
            set { Interface.SetValue(id, "deviation", value); }
        }

        public double Variation
        {
            get { return Interface.GetValue(id, "variation"); }
            set { Interface.SetValue(id, "variation", value); }
        }

        public double Location
        {
            get { return Interface.GetValue(id, "location"); }
            set { Interface.SetValue(id, "location", value); }
        }

        public double Scale
        {
            get { return Interface.GetValue(id, "scale"); }
            set { Interface.SetValue(id, "scale", value); }
        }

        public double Shift
        {
            get { return Interface.GetValue(id, "shift"); }
            set { Interface.SetValue(id, "set_shift", value); }
        }

        public void SetShift(double value)
        {
            Interface.SetValue(id, "shift", value);
        }

        public double ShiftB
        {
            get { return Interface.GetValue(id, "shift_b"); }
            set { Interface.SetValue(id, "shift_b", value); }
        }

        public double Minimum
        {
            get { return Interface.GetValue(id, "minimum"); }
            set { Interface.SetValue(id, "minimum", value); }
        }

        public double Maximum
        {
            get { return Interface.GetValue(id, "maximum"); }
            set { Interface.SetValue(id, "maximum", value); }
        }

        public double Shape
        {
            get { return Interface.GetValue(id, "shape"); }
            set { Interface.SetValue(id, "shape", value); }
        }

        public double ShapeB
        {
            get { return Interface.GetValue(id, "shape_b"); }
            set { Interface.SetValue(id, "shape_b", value); }
        }

        public int Observations
        {
            get { return Interface.GetIntValue(id, "observations"); }
            set { Interface.SetIntValue(id, "observations", value); }
        }

        public IList<HistogramValue> HistogramValues
        {
            get
            {
                if (histogramValues == null)
                {
                    histogramValues = new CallBackList<HistogramValue>(HistogramValuesChanged);

                    int[] histogramIds = Interface.GetArrayIdValue(id, "histogram_values");
                    foreach (int histogramId in histogramIds)
                    {
                        histogramValues.AddWithoutCallBack(new HistogramValue(histogramId));
                    }
                }

                return histogramValues;
            }
        }

        private void HistogramValuesChanged(ListOperationType listOperation, HistogramValue item)
        {
            Interface.SetArrayIntValue(id, "histogram_values", this.histogramValues.Select(p => p.GetId()).ToArray());
        }

        public IList<DiscreteValue> DiscreteValues
        {
            get
            {
                if (discreteValues == null)
                {
                    discreteValues = new CallBackList<DiscreteValue>(DiscreteValuesChanged);

                    int[] discreteIds = Interface.GetArrayIdValue(id, "discrete_values");
                    foreach (int discreteId in discreteIds)
                    {
                        discreteValues.AddWithoutCallBack(new DiscreteValue(discreteId));
                    }
                }

                return discreteValues;
            }
        }

        private void DiscreteValuesChanged(ListOperationType listOperation, DiscreteValue item)
        {
            Interface.SetArrayIntValue(id, "discrete_values", this.discreteValues.Select(p => p.GetId()).ToArray());
        }

        public IList<FragilityValue> FragilityValues
        {
            get
            {
                if (fragilityValues == null)
                {
                    fragilityValues = new CallBackList<FragilityValue>(FragilityValuesChanged);

                    int[] fragilityIds = Interface.GetArrayIdValue(id, "fragility_values");
                    foreach (int fragilityId in fragilityIds)
                    {
                        fragilityValues.AddWithoutCallBack(new FragilityValue(fragilityId));
                    }
                }

                return fragilityValues;
            }
        }

        private void FragilityValuesChanged(ListOperationType listOperation, FragilityValue item)
        {
            if (listOperation == ListOperationType.Add)
            {
                Interface.SetIntValue(id, "fragility_values", item.GetId());
            }
            else
            {
                Interface.SetArrayIntValue(id, "fragility_values", this.fragilityValues.Select(p => p.GetId()).ToArray());
            }
        }

        public IList<ContributingStochast> ContributingStochasts
        {
            get
            {
                if (contributingStochasts == null)
                {
                    contributingStochasts = new CallBackList<ContributingStochast>(ContributingStochastsChanged);

                    int[] stochastIds = Interface.GetArrayIdValue(id, "contributing_stochasts");
                    foreach (int stochastId in stochastIds)
                    {
                        contributingStochasts.AddWithoutCallBack(new ContributingStochast(stochastId));
                    }
                }

                return contributingStochasts;
            }
        }

        private void ContributingStochastsChanged(ListOperationType listOperation, ContributingStochast item)
        {
            Interface.SetArrayIntValue(id, "contributing_stochasts", this.contributingStochasts.Select(s => s.GetId()).ToArray());
        }

        public double DesignQuantile
        {
            get { return Interface.GetValue(id, "design_quantile"); }
            set { Interface.SetValue(id, "design_quantile", value); }
        }

        public double DesignFactor
        {
            get { return Interface.GetValue(id, "design_factor"); }
            set { Interface.SetValue(id, "design_factor", value); }
        }

        public double DesignValue
        {
            get { return Interface.GetValue(id, "design_value"); }
            set { Interface.SetValue(id, "design_value", value); }
        }

        public bool IsArray
        {
            get { return Interface.GetBoolValue(id, "is_array"); }
            set { Interface.SetBoolValue(id, "is_array", value); }
        }

        public int ArraySize
        {
            get { return Interface.GetIntValue(id, "array_size"); }
            set { Interface.SetIntValue(id, "array_size", value); }
        }

        public double GetQuantile(double quantile)
        {
            return Interface.GetArgValue(id, "quantile", quantile);
        }

        public double GetXFromU(double u)
        {
            return Interface.GetArgValue(id, "x_from_u", u);
        }

        public virtual double GetUFromX(double x)
        {
            return Interface.GetArgValue(id, "u_from_x", x);
        }

        public double GetXFromP(double p)
        {
            return Interface.GetArgValue(id, "x_from_p", p);
        }

        public double GetPFromX(double p)
        {
            return Interface.GetArgValue(id, "p_from_x", p);
        }

        public double GetPDF(double x)
        {
            return Interface.GetArgValue(id, "pdf", x);
        }

        public double GetCDF(double x)
        {
            return Interface.GetArgValue(id, "cdf", x);
        }

        public double GetUFromXAndSource(double xSource, double x)
        {
            Interface.SetArrayValue(id, "x_and_source", [x, xSource]);
            Interface.Execute(id, "initialize_conditional_values");
            return Interface.GetValue(id, "u_from_x_and_source");
        }

        public double GetXFromUAndSource(double u, double x)
        {
            Interface.SetArrayValue(id, "u_and_x", [u, x]);
            Interface.Execute(id, "initialize_conditional_values");
            return Interface.GetValue(id, "x_from_u_and_source");
        }

        public void SetXAtU(double x, double u, ConstantParameterType constantParameterType)
        {
            Interface.SetArrayValue(id, "u_and_x", [u, x]);
            if (constantParameterType == ConstantParameterType.Deviation)
            {
                Interface.Execute(id, "set_x_at_u_dev");
            }
            else if (constantParameterType == ConstantParameterType.VariationCoefficient)
            {
                Interface.Execute(id, "set_x_at_u_var");
            }
        }

        public bool IsVarying()
        {
            return Interface.GetBoolValue(id, "is_varying");
        }

        public bool IsConditional
        {
            get { return Interface.GetBoolValue(id, "conditional"); }
            set { Interface.SetBoolValue(id, "conditional", value); }
        }

        public Stochast ConditionalSource
        {
            get
            {
                if (conditionalSource == null)
                {
                    int sourceId = Interface.GetIdValue(id, "conditional_source");
                    if (sourceId > 0)
                    {
                        conditionalSource = ObjectFactory.GetObject<Stochast>(sourceId);
                    }
                }

                return conditionalSource;
            }
            set
            {
                if (value != null)
                {
                    Interface.SetIntValue(id, "conditional_source", value.GetId());
                }

                conditionalSource = value;
            }
        }

        public IList<ConditionalValue> ConditionalValues
        {
            get
            {
                if (conditionalValues == null)
                {
                    conditionalValues = new CallBackList<ConditionalValue>(ConditionalValuesChanged);

                    int[] conditionalValueIds = Interface.GetArrayIdValue(id, "conditional_values");
                    foreach (int conditionalValueId in conditionalValueIds)
                    {
                        conditionalValues.AddWithoutCallBack(new ConditionalValue(conditionalValueId));
                    }
                }

                return conditionalValues;
            }
        }

        private void ConditionalValuesChanged(ListOperationType listOperation, ConditionalValue item)
        {
            Interface.SetArrayIntValue(id, "conditional_values", this.conditionalValues.Select(s => s.GetId()).ToArray());
        }

        public Stochast GetConditionalStochast(double x)
        {
            Interface.SetValue(id, "conditional_x", x);
            int stochastId = Interface.GetIdValue(id, "conditional_x");

            return new Stochast(stochastId);
        }

        public bool CanFit()
        {
            return Interface.GetBoolValue(id, "can_fit");
        }

        public void Fit(double[] values, double shift = double.NaN)
        {
            Interface.SetIntValue(id, "prior", 0);
            Interface.SetValue(id, "shift_for_fit", shift);
            Interface.SetArrayValue(id, "data", values);

            Interface.Execute(id, "fit");

            this.histogramValues = null;
            this.discreteValues = null;
            this.fragilityValues = null;
        }

        public bool CanFitPrior()
        {
            return Interface.GetBoolValue(id, "can_fit_prior");
        }

        public void FitPrior(double[] values, Stochast prior, double shift = double.NaN)
        {
            Interface.SetIntValue(id, "prior", prior?.GetId() ?? 0);
            Interface.SetValue(id, "shift_for_fit", shift);
            Interface.SetArrayValue(id, "data", values);

            Interface.Execute(id, "fit_prior");

            this.histogramValues = null;
            this.discreteValues = null;
            this.fragilityValues = null;
        }

        public void FitWeighted(double[] values, double[] weights)
        {
            Interface.SetIntValue(id, "prior", 0);
            Interface.SetValue(id, "shift_for_fit", 0);
            Interface.SetArrayValue(id, "data", values);
            Interface.SetArrayValue(id, "weights", values);

            Interface.Execute(id, "fit_weighted");

            this.histogramValues = null;
            this.discreteValues = null;
            this.fragilityValues = null;
        }

        public double GetKSTest(double[] values)
        {
            Interface.SetArrayValue(id, "data", values);
            return Interface.GetValue(id, "ks_test");
        }

        public double[] GetSpecialXValues()
        {
            return Interface.GetArrayValue(id, "special_values");
        }

        public void CopyFrom(Stochast source)
        {
            if (source != null)
            {
                Interface.SetIntValue(id, "copy_from", source.GetId());

                this.histogramValues = null;
                this.discreteValues = null;
                this.fragilityValues = null;
                this.conditionalValues = null;
                this.contributingStochasts = null;
            }
        }

        public void InitializeForRun()
        {
            Interface.Execute(id, "initialize_for_run");
        }

        public bool CanTruncate()
        {
            return Interface.GetBoolValue(id, "can_truncate");
        }

        public bool CanInvert()
        {
            return Interface.GetBoolValue(id, "can_invert");
        }

        public bool HasParameter(DistributionPropertyType property)
        {
            switch (property)
            {
                case DistributionPropertyType.Location: return Interface.GetBoolValue(id, "is_used_location");
                case DistributionPropertyType.Scale: return Interface.GetBoolValue(id, "is_used_scale");
                case DistributionPropertyType.Minimum: return Interface.GetBoolValue(id, "is_used_minimum");
                case DistributionPropertyType.Maximum: return Interface.GetBoolValue(id, "is_used_maximum");
                case DistributionPropertyType.Shift: return Interface.GetBoolValue(id, "is_used_shift");
                case DistributionPropertyType.ShiftB: return Interface.GetBoolValue(id, "is_used_shift_b");
                case DistributionPropertyType.Shape: return Interface.GetBoolValue(id, "is_used_shape");
                case DistributionPropertyType.ShapeB: return Interface.GetBoolValue(id, "is_used_shape_b");
                case DistributionPropertyType.Observations: return Interface.GetBoolValue(id, "is_used_observations");
                default: throw new System.NotSupportedException(property.ToString());
            }
        }

        public bool IsQualitative()
        {
            return Interface.GetBoolValue(id, "is_qualitative");
        }

        public bool IsValid()
        {
            return Interface.GetBoolValue(id, "is_valid");
        }
    }
}
